### PR TITLE
Add Vagrantbackend and OSX Support

### PR DESCRIPTION
**Pull request by codylane**
Hello -

Thanks again for creating such an awesome framework, I hope that these new additions are acceptable and will be merged into master.

Summary Of Changes
================
1. New VagrantBackend to more easily work with Vagrant environments.  The original plan for this backend was to be used by testinfra and spin up MacOS vagrant boxes (setup and teardown) when vagrant tests are specified using the 'vagrant_sut' pytest fixture.
2.  A few new enhancements to `system_info` to reduce redundant operating system testing.
3.  A new homebrew package implementation for OSX based hosts.
4.  A new LaunchD service implementation for OSX based hosts.
5.  Flake8 and Pylint fixes to all of the framework.
6.  Small refactors in some major components of testinfra to make readability a little easier IMHO.

Things I Still Need To Do
==================
I plan on tackling these in the next day or two but wanted to get this pull request in ASAP.

- [ ] I still need to add some tests for LaunchDService.
- [ ] I still need to add tests for system_info enhancements.

 I'm happy to refactor, explain or fix any constructive suggestions you may have so that this will be merged into master.  This has been something I've been working on for the past few months and would love some feedback.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,15 +1,55 @@
...
+      env: TOXENV=flake8
+    - os: osx
+      osx_image: xcode8.3
+      language: generic
+      env: TOXENV=pylint
+    - os: osx
+      osx_image: xcode8.3
+      language: generic
+      env: TOXENV=sphinxdoc
+
```
Comment (philpep) [inline .travis.yml]: I'm not sure we want sphinx, pylint, flake8 tested on OSX. Having only testinfra OSX specific tests should be enough.

File: testinfra/backend/vagrant.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        cmd.add('is_not_created', not created)
+
+        self._status = cmd
+        return self._status
+
+    def _refresh_ssh_config(self):
+        if self._ssh_config is None:
+            self._ssh_config = self.run_vagrant('vagrant ssh-config %s',
+                                                self.box
+                                                )
```
Comment (philpep) [inline testinfra/backend/vagrant.py]: I'm not sure what's the benefit of having a (very complex) vagrant backend over integrate vagrant and testinfra in a shell script ?

```
!/bin/sh
vagrant up
vagrant ssh-config > ssh_config
testinfra --ssh-config ssh_config --host user@box
vagrant stop
```

File: images/centos_6/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    ssh-keygen -q -N "" -t rsa -f /etc/ssh/ssh_host_rsa_key && \
+    sed -i "s/#UsePrivilegeSeparation.*/UsePrivilegeSeparation no/g" /etc/ssh/sshd_config && \
+    chkconfig sshd on && \
+    service sshd start && \
+    mkdir -p /root/.ssh && \
+    echo "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCgDryK4AjJeifuc2N54St13KMNlnGLAtibQSMmvSyrhH7XJ1atnBo1HrJhGZNNBVKM67+zYNc9J3fg3qI1g63vSQAA+nXMsDYwu4BPwupakpwJELcGZJxsUGzjGVotVpqPIX5nW8NBGvkVuObI4UELOleq5mQMTGerJO64KkSVi20FDwPJn3q8GG2zk3pESiDA5ShEyFhYC8vOLfSSYD0LYmShAVGCLEgiNb+OXQL6ZRvzqfFEzL0QvaI/l3mb6b0VFPAO4QWOL0xj3cWzOZXOqht3V85CZvSk8ISdNgwCjXLZsPeaYL/toHNvBF30VMrDZ7w4SDU0ZZLEsc/ezxjb" > /root/.ssh/authorized_keys
+
+VOLUME ["/sys/fs/cgroup"]
+EXPOSE 22
+CMD ["/usr/sbin/sshd", "-D"]
```
Comment (philpep) [inline images/centos_6/Dockerfile]: Why adding centos 6 tests ?

File: setup.cfg [hunk 4]
```diff
@@ -38,6 +38,9 @@ console_scripts =
 pytest11 =
     pytest11.testinfra = testinfra.plugin
 
+[flake8]
+ignore = E501
+
```
Comment (philpep) [inline setup.cfg]: Until now we follow E501, let's keep this.

File: test/conftest.py [hunk 5]
```diff
@@ -135,76 +189,203 @@ def teardown():
...
+        return host, mocked_method
+
+    return on_call
+
+
 @pytest.fixture
 def host(request, tmpdir_factory):
     if not has_docker():
-        pytest.skip()
+        pytest.skip('Skipping test because docker is not installed')
```
Comment (philpep) [inline test/conftest.py]: This hunk is useful and could be merged right now if in a dedicated commit.

File: test/test_modules.py [hunk 6]
```diff
@@ -220,7 +242,11 @@ def test_nonexistent_user(host):
 def test_current_user(host):
     assert host.user().name == "root"
     pw = host.user().password
-    assert crypt.crypt("foo", pw) == pw
+
+    if HAS_PASSLIB:
+        assert passlib.hash.sha512_crypt.verify('foo', pw)
```
Comment (philpep) [inline test/test_modules.py]: Why using passlib over crypt ?

File: test/test_modules.py [hunk 7]
```diff
@@ -371,7 +397,7 @@ def test_supervisor(host):
     host.run("service supervisor stop")
     assert not host.service("supervisor").is_running
     with pytest.raises(RuntimeError) as excinfo:
-        host.supervisor("tail").is_running
+        assert host.supervisor("tail").is_running
```
Comment (philpep) [inline test/test_modules.py]: Why this is needed ? pytest.raises will raise if RuntimeError is not raised.

File: testinfra/main.py [hunk 8]
```diff
@@ -52,11 +52,12 @@ def report(self):
...
-            b"TESTINFRA %s - %d passed, %d failed, %d skipped in %.2f "
-            b"seconds\n") % (
-                status, self.passed, self.failed, self.skipped, self.total_time
-            ))
+        sys.stdout.write(
+            (
+                b"TESTINFRA %s - %d passed, %d failed, %d skipped in %.2f "
+                b"seconds\n"
+            ) % (status, self.passed, self.failed, self.skipped, self.total_time)
+        )
```
Comment (philpep) [inline testinfra/main.py]: unrelated hunk

File: testinfra/modules/service.py [hunk 9]
```diff
@@ -46,21 +80,19 @@ def is_enabled(self):
...
+            bsd_service_provider,
+            darwin_service_provider,
+        ]
+
+        in_service_providers = [provider(host) for provider in providerz]
+
+        svc_provider = [provider for provider in in_service_providers if provider is not None]
+        if svc_provider:
+            return svc_provider[0]
+
```
Comment (philpep) [inline testinfra/modules/service.py]: I think previous version of the code was more readable

**Comment by codylane** [comment]
Hi @philpep - Thank you for your feedback and for your time reviewing.   I agree that it's a large pull request and I tried my best to keep everything related and to the point.  A month or so I was planning on separating each component out and then I went on vacation and then attempted to pick up where I left off.

I've wondered for a while now if the VagrantBackend makes sense because my original intention was for it to be used as my setup and teardown for firing up a vagrant box, doing some testing, then shutting down the vagrant box.  I had all this working then ripped out most of it when I got back from vacation because it bloated and Travis testing was very, very slow and I might be a tad impatient. 

I'll go through your comments a little later and replay and fix where I can.  Thank you for your response.

Lastly, do you recommend we close this pull request in favor me opening pull requests for each one of the following?
* Breaking VagrantBackend only it's code + tests into it's own PR? - Still deciding as you suggested if this even makes sense.
* Breaking out the HomebrewPackage + tests into it's own PR?
* Breaking out the LaunchDService + tests into it's own PR?
* Breaking out the system_info + tests enhances into it's own PR?

File: testinfra/backend/vagrant.py (hunk 2 above)
Comment (codylane) [inline testinfra/backend/vagrant.py]: My original intention with this was for testinfra to do the setup:
1. Start the Vagrant box
2. Prep the vagrant box with needed libraries so that our tests will run
3. Provide way to simplify creating the hooks that could be used during your CI testing and using Vagrant through testinfra where you focus on the outcome of tests rather than fussing around with getting vagrant to work. 
4. Testing shell scripts is hard, testing python is much easier and this library is tested and it should work with the latest versions of Vagrant and I totally plan on helping maintain it if you find this library helpful for testinfra.
5. This backend was primarily created to be used as a test fixture for (spin up, run your tests, tear it down) and repeat and only focusing on solid building blocks.

You won't hurt my feelings if you say you don't think it will be a good fit for this project.  Either way, I'm going to keep this code around in my repo incase others find it useful.

File: testinfra/main.py (hunk 8 above)
Comment (codylane) [inline testinfra/main.py]: Sorry, my commit history got whacked when I was trying to deal with all the merge conflicts when I got back from vacation.  The context here is that the new additions above actually pass flake8 and pylint tests. They wouldn't until I made those changes.

File: .travis.yml (hunk 1 above)
Comment (codylane) [inline .travis.yml]: Fair enough, was just trying to stick with your original implementation.  I'll remove this.

File: setup.cfg (hunk 4 above)
Comment (codylane) [inline setup.cfg]: K.  I'll also remove this.  Thanks for your response.

Comment (codylane) [inline setup.cfg]: I have a feeling there are going to a be a lot of failed pylint tests after undoing this and attempting to fix those lines will make the code really ugly since 80 chars isn't really valid anymore on normal terminals.  Just my 3$.

File: images/centos_6/Dockerfile (hunk 3 above)
Comment (codylane) [inline images/centos_6/Dockerfile]: I did this to increase test coverage for missing tests.  As an example, non of the current docker images in this repo actually tested the 'SysvService' in testinfra/modules/servie.py, it does now with this docker file.
